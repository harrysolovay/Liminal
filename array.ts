import { make, type Ty } from "./_base.ts"

export function array<E extends Ty>(description: string | undefined, element: E): ArrayTy<E> {
  return make((ctx) => ({
    type: "array",
    description,
    items: ctx.ref(element),
  }))
}

export type ArrayTy<E extends Ty = any> = Ty<Array<InstanceType<E>>>
