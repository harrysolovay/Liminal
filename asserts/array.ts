import { assert } from "../util/mod.ts"

export function minLength<T>(minLength: number, value: Array<T>) {
  assert(value.length >= minLength, `Length must be gte ${minLength} elements.`)
}

export function maxLength<T>(maxLength: number, value: Array<T>) {
  assert(value.length <= maxLength, `Length must be lte ${maxLength} elements.`)
}
