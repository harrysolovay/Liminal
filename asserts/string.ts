import { assert } from "../util/mod.ts"

export function minLength(value: string, minLength: number) {
  assert(value.length >= minLength, `Length must be gte ${minLength} chars.`)
}

export function maxLength(value: string, maxLength: number) {
  assert(value.length <= maxLength, `Length must be lte ${maxLength} chars.`)
}
