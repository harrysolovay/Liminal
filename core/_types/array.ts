import type { Type } from "../Type.ts"
import { declareType } from "./_declareType.ts"

export function array<T>(element: Type<T>): Type<Array<T>> {
  return declareType(() => array<T>, [element])
}
