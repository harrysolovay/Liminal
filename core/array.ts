import type { Type } from "../Type.ts"
import { declare } from "../TypeDeclaration.ts"
import { assert } from "../util/assert.ts"

export function array<E extends Type>(element: E): Type<Array<E["T"]>, {
  minLength: number
  maxLength: number
}, E["P"]> {
  return declare<Array<E["T"]>>()({
    name: "array",
    source: {
      factory: array,
      args: { element },
    },
    subschema: (visit) => ({
      type: "array",
      items: visit(element),
    }),
    process: (value, visit) => value.map((v, i) => visit(v, element, i)),
    assertRefinementsValid: ({ minLength, maxLength }) => {
      assert(
        !(typeof minLength === "number" && typeof maxLength === "number") || minLength <= maxLength,
      )
    },
    assertRefinements: {
      minLength: (value, minLength) => value.length >= minLength,
      maxLength: (value, maxLength) => value.length <= maxLength,
    },
  })
}
