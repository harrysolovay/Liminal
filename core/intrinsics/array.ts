import { declareType } from "../declareType.ts"
import type { Type } from "../Type.ts"

export function array<T, E>(element: Type<T, E>): Type<Array<T>, E> {
  return declareType({
    type: "array",
    self() {
      return array
    },
    args: [element],
  })
}
