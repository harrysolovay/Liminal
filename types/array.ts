import { type AnyType, declareType, type Type } from "../core/mod.ts"

export function array<E extends AnyType>(element: E): Type<Array<E["T"]>, E["P"]> {
  return declareType({
    name: "array",
    source: {
      factory: array,
      args: [element],
    },
    visitValue: (value, visit) => value.map((e, i) => visit(e, element, i)),
  })
}
