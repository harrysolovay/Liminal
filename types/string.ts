import { declare, type Type } from "../core/mod.ts"
import { assert } from "../util/assert.ts"

export const string: Type.Initial<string, {
  minLength: number
  maxLength: number
}> = declare({
  name: "string",
  source: {
    getType: () => string,
  },
  refinements: ({ minLength, maxLength }) => {
    assert(
      !(typeof minLength === "number" && typeof maxLength === "number") || minLength <= maxLength,
      "`minLength` cannot be greater than `maxLength`.",
    )
    return {
      minLength: `Length must be gte ${minLength} chars.`,
      maxLength: `Length must be lte ${maxLength} chars.`,
    }
  },
  subschema: () => ({ type: "string" }),
  output: (f) =>
    f({
      refinementPredicates: {
        minLength: (value, minLength) => value.length >= minLength,
        maxLength: (value, maxLength) => value.length <= maxLength,
      },
    }),
})
