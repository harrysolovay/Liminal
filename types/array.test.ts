import { assertTypeSnapshot } from "testing"
import { array } from "./array.ts"
import { string } from "./string.ts"

Deno.test("array", async (t) => {
  await assertTypeSnapshot(t, array(string))
})
