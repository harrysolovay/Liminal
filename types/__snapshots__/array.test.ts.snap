export const snapshot = {};

snapshot[`array 1`] = `
{
  items: {
    type: "string",
  },
  type: "array",
}
`;

snapshot[`array 2`] = `
{
  description: "One.",
  items: {
    type: "string",
  },
  type: "array",
}
`;

snapshot[`array 3`] = `
{
  description: "One. Two.",
  items: {
    type: "string",
  },
  type: "array",
}
`;
